discretisedfield/region.py

```diff
@@ -172,7 +222,148 @@ def pmax(self):
         """
         return self._pmax
 
-    @functools.cached_property
+    @pmax.setter
+    def pmax(self, value):
+        if not isinstance(value, (tuple, list, np.ndarray)):
+            raise TypeError(
+                "pmax must be of type tuple, list, or np.ndarray. Not"
+                f" type(pmax)={type(value)}."
+            )
+        if not all(isinstance(i, numbers.Number) for i in value):
+            raise TypeError("pmax can only contain elements of type numbers.Number.")
+
+        if not len(value) == self.ndim:
+            raise ValueError(
+                "The length of pmax must be the same as the number of"
+                f" dimensions: {self.ndim=}. Not len(pmax)={len(value)}."
+            )
+
+        old_pmax = self._pmax
+        self._pmax = np.array(value)
+        if not np.all(self.edges):
+            edges = self.edges
+            self._pmax = old_pmax
+            raise ValueError(
+                f"At least one of the region's edge lengths is zero: {edges}."
+            )
+
+    @property
```

Comment:
I think this can't change after we create the region. So we could use `functools.cached_property`.